chain/rust/src/utils.rs

```diff
@@ -569,6 +569,302 @@ impl Deserialize for NetworkId {
     }
 }
 
+// Represents the cddl: #6.258([+ T]) / [* T]
+// it DOES NOT and CAN NOT have any encoding detials per element!
+// so you can NOT use it on any primitives so must be serializable directly
+#[derive(Debug, Clone)]
+pub struct NonemptySet<T> {
+    elems: Vec<T>,
+    len_encoding: LenEncoding,
+    // also controls whether to use the tag encoding (Some) or raw array (None)
+    tag_encoding: Option<Sz>,
+}
+
+impl<T: serde::Serialize> serde::Serialize for NonemptySet<T> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        self.elems.serialize(serializer)
+    }
+}
+
+impl<'de, T: serde::de::Deserialize<'de>> serde::de::Deserialize<'de> for NonemptySet<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::de::Deserializer<'de>,
+    {
+        Vec::deserialize(deserializer).map(|elems| Self {
+            elems,
+            len_encoding: LenEncoding::default(),
+            tag_encoding: None,
+        })
+    }
+}
+
+impl<T: schemars::JsonSchema> schemars::JsonSchema for NonemptySet<T> {
+    fn schema_name() -> String {
+        Vec::<T>::schema_name()
+    }
+    fn json_schema(gen: &mut schemars::gen::SchemaGenerator) -> schemars::schema::Schema {
+        Vec::<T>::json_schema(gen)
+    }
+    fn is_referenceable() -> bool {
+        Vec::<T>::is_referenceable()
+    }
+}
+
+impl<T> AsRef<[T]> for NonemptySet<T> {
```

Comment:
would be nice to also have `IntoIterator for &'a NonemptySet<T>` and I guess `NonemptySet<T>` I think. For loop syntax is quite convenient.

Reply:
Is there anything else we need or should `Deref` and `DerefMut` take care of that?